numpy-stubs/__init__.pyi

```diff
@@ -792,3 +797,124 @@ class AxisError(ValueError, IndexError):
     def __init__(
         self, axis: int, ndim: Optional[int] = ..., msg_prefix: Optional[str] = ...
     ) -> None: ...
+
+# Functions from np.core.fromnumeric
+_Mode = Literal["raise", "wrap", "clip"]
+_Order = Literal["C", "F", "A"]
+_PartitionKind = Literal["introselect"]
+_SortKind = Literal["quicksort", "mergesort", "heapsort", "stable"]
+
+# Various annotations for scalars
+_ScalarGeneric = TypeVar("_ScalarGeneric", bound=generic)
+_ScalarBuiltin = Union[str, bytes, dt.date, dt.timedelta, bool, int, float, complex]
+_Scalar = Union[_ScalarBuiltin, generic]
+
+# An array-like object consisting of integers
+_Int = Union[int, integer]
+_ArrayLikeIntNested = Any  # TODO: wait for support for recursive types
+_ArrayLikeInt = Union[
+    _Int, ndarray, Sequence[_Int], Sequence[_ArrayLikeIntNested]  # TODO: ndarray[int]
+]
+
+# An array-like object consisting of strings
+_ArrayLikeStr = Union[
+    str, str_, Sequence[Union[str, str_]], ndarray
+]  # TODO: ndarray[str]
+
+# The signature of take() follows a common theme with its overloads:
+# 1. A generic comes in; the same generic comes out
+# 2. A scalar comes in; a generic comes out
+# 3. An array-like object comes in; some keyword ensures that a generic comes out
+# 4. An array-like object comes in; an ndarray or generic comes out
+@overload
+def take(
+    a: _ScalarGeneric,
+    indices: int,
+    axis: Optional[int] = ...,
+    out: Optional[ndarray] = ...,
+    mode: _Mode = ...,
+) -> _ScalarGeneric: ...
+@overload
+def take(
+    a: _Scalar,
+    indices: int,
+    axis: Optional[int] = ...,
+    out: Optional[ndarray] = ...,
+    mode: _Mode = ...,
+) -> generic: ...
```

Comment:
This isn't always true I don't think:

```python
>>> type(np.take(datetime.timedelta(days=1), 0))
<class 'datetime.timedelta'>
>>> type(np.take(datetime.date.today(), 0))
<class 'datetime.date'>
```

Works for the others though. There might be diminishing returns getting this to work until such a time as we have a templating system in place; then it can generate specific overloads like `bool -> np.bool_` etc.

Reply:
If thus is only exception to the rule, then I feel it might be worthwhile to swap ``generic`` with an union for the time being.